okio/src/main/java/okio/TeeSource.java

```diff
@@ -0,0 +1,41 @@
+package okio;
+
+import java.io.IOException;
+
+
+class TeeSource implements Source {
+
+  private final Source source;
+  private final BufferedSink cacheBody;
+
+  public TeeSource(Source source, BufferedSink cacheBody) {
+    this.source = source;
+    this.cacheBody = cacheBody;
+  }
+
+  @Override
+  public long read(Buffer sink, long byteCount) throws IOException {
+    long bytesRead;
+    bytesRead = source.read(sink, byteCount);
+
+    if (bytesRead == -1) {
+      cacheBody.close(); // The cache response is complete!
```

Comment:
make sense removing sink close from feature.  thank you.

On Fri, Jan 29, 2016 at 6:09 PM, Jake Wharton <notifications@github.com>
wrote:

> In okio/src/main/java/okio/TeeSource.java
> <https://github.com/square/okio/pull/187#discussion_r51328542>:
>
> > +
> > +  private final Source source;
> > +  private final BufferedSink cacheBody;
> > +
> > +  public TeeSource(Source source, BufferedSink cacheBody) {
> > +    this.source = source;
> > +    this.cacheBody = cacheBody;
> > +  }
> > +
> > +  @Override
> > +  public long read(Buffer sink, long byteCount) throws IOException {
> > +    long bytesRead;
> > +    bytesRead = source.read(sink, byteCount);
> > +
> > +    if (bytesRead == -1) {
> > +      cacheBody.close(); // The cache response is complete!
>
> You don't think it's surprising behavior to close the tee sink when
> application code never called close on this source? Not calling close on
> the source would be a bug (and potential leak).
>
> For OkHttp's use, we can implement this early auto-close behavior on top
> of this Okio feature.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/square/okio/pull/187/files#r51328542>.
>